unittests/test-contracts/if_ibc/if_ibc.hpp

```diff
@@ -0,0 +1,489 @@
+
+      static uint32_t reverse_bytes(const uint32_t input){
+         uint32_t output = (input>>24 & 0xff)|(input>>8 & 0xff00)|(input<<8 & 0xff0000)|(input<<24 & 0xff000000);
+         return output;
+      }
+
+      static checksum256 hash_pair(const std::pair<checksum256, checksum256> p){
+         std::array<uint8_t, 32> arr1 = p.first.extract_as_byte_array();
+         std::array<uint8_t, 32> arr2 = p.second.extract_as_byte_array();
+         std::array<uint8_t, 64> result;
+         std::copy (arr1.cbegin(), arr1.cend(), result.begin());
+         std::copy (arr2.cbegin(), arr2.cend(), result.begin() + 32);
+         checksum256 hash = sha256(reinterpret_cast<char*>(result.data()), 64);
+         return hash;
+      }
+
+      static time_point add_time(const time_point& time, const uint32_t seconds ){
+         int64_t total_seconds = (static_cast<int64_t>(time.sec_since_epoch()) + static_cast<int64_t>(seconds));
+         microseconds ms = microseconds(total_seconds * 1000000);
+         time_point tp = time_point(ms);
+         return tp;
+      }
+
+      /*
+
+      //discuss : compute merkle branch direction vs providing them as part of the proof
+
+      static std::vector<uint8_t> _get_directions(const uint64_t index, const uint64_t last_node_index){
+
+          std::vector<uint8_t> proof;
+
+          uint64_t c_index = index;
+          uint64_t layers_depth = calculate_max_depth(last_node_index) -1;
+          uint64_t c_last_node_index = last_node_index;
+
+          for (uint64_t i = 0; i < layers_depth; i++) {
+              if (c_last_node_index % 2) c_last_node_index+=1;
+
+              bool isLeft = c_index % 2 == 0 ? 0 : 1;
+              uint64_t pairIndex = isLeft ? c_index - 1 :
+                          c_index == last_node_index - 1 && i < layers_depth - 1 ? c_index :
+                          c_index + 1;
+
+              c_last_node_index/=2;
+              if (pairIndex < last_node_index) proof.push_back(isLeft);
+
+              c_index = c_index / 2;
+
+          }
+
+          return proof;
+
+      }
+
+      */
+
+      struct merkle_branch {
+
+         uint8_t direction;
+         checksum256 hash;
+
+      };
+
+
+      //compute the merkle root of target node and vector of merkle branches
+      static checksum256 _compute_root(const std::vector<merkle_branch> proof_nodes, const checksum256& target){
+
+          checksum256 hash = target;
+
+          for (int i = 0 ; i < proof_nodes.size() ; i++){
+              const checksum256 node = proof_nodes[i].hash;
+              std::array<uint8_t, 32> arr = node.extract_as_byte_array();
+
+              if (proof_nodes[i].direction == 0){
+                  hash = hash_pair(std::make_pair(hash, node));
+              }
+              else {
+                  hash = hash_pair(std::make_pair(node, hash));
+              }
+
+          }
+
+          return hash;
+
+      }
+
+      struct quorum_certificate {
+          std::vector<uint64_t> finalizers;

```

Comment:
Yes, this is the underlying data type of the bitset. Since the quorum_certificate is part of the data structure that must be converted to and from binary via the ABI, I don't think we can pass the bitset type directly, unless we make it an ABI type. I'm not sure if this is desirable. But I can try if this is preferred.